cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+func upgrade(rootArgs *rootArgs, args *upgradeArgs) (err error) {
+	initLogsOrExit(rootArgs)
+	l.logAndPrintf("Client - istioctl version: %s\n", opversion.OperatorVersionString)
+
+	targetValues := genValuesFromFile(args.inFilename, args.force)
+	targetICPS := genICPSFromFile(args.inFilename, args.force)
+	targetVersion := targetICPS.GetTag()
+	l.logAndPrintf("Upgrade - target version: %s\n", targetVersion)
+
+	kubeClient := getKubeExecClient(args.kubeConfigPath, args.context)
+	//TODO(elfinhe): support components distributed in multiple namespaces
+	istioNamespace := targetICPS.GetDefaultNamespace()
+	currentVer := retrieveControlPlaneVersion(kubeClient, istioNamespace)
+	currentValues := readValuesFromInjectorConfigMap(kubeClient, istioNamespace)
+
+	if !args.force {
+		checkSupportedVersions(currentVer, targetVersion, args.versionsURI)
+		checkUpgradeValues(currentValues, targetValues, args.yes)
+	}
+
+	runPreUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun)
+	applyUpgradeManifest(args.inFilename, args.kubeConfigPath,
+		args.context, rootArgs.dryRun, rootArgs.verbose)
+	runPostUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun)
+
+	if args.wait {
+		waitUpgradeComplete(kubeClient, istioNamespace, targetVersion)
+		upgradeVer := retrieveControlPlaneVersion(kubeClient, istioNamespace)
+		l.logAndPrintf("Success. Now the Istio control plane is running at version %v.", upgradeVer)
+		return
+	}
+	l.logAndPrintf("Upgrade submitted. Please use `istioctl version` to check the current versions.")
+	return
+}
+
+// applyUpgradeManifest applies the Istion Control Plane specs reading from inFilename to
+// the cluster by given kubeConfigPath and context
+func applyUpgradeManifest(inFilename, kubeConfigPath, context string, dryRun, verbose bool) {
+	genApplyManifests(nil, inFilename, dryRun,
+		verbose, kubeConfigPath, context, upgradeWaitSecWhenApply, l)
+}
+
+// checkUpgradeValues checks the upgrade eligibility by comparing the current values with the target values
+func checkUpgradeValues(curValues string, tarValues string, yes bool) {
+	diff := compare.YAMLCmp(curValues, tarValues)
+	if diff == "" {
+		l.logAndPrintf("Upgrade check: Values unchanged. The target values are identical to the current values.\n")
+	} else {
+		l.logAndPrintf("Upgrade check: Warning!!! the following values will be changed as part of upgrade. "+
+			"If you have not overridden these values, they will change in your cluster. Please double check they are correct:\n%s", diff)
+	}
+
+	if yes {
+		return
+	}
+	if !confirm("Confirm to proceed [y/N]?", os.Stdout) {
+		l.logAndFatalf("Abort.")
+	}
+}
+
+// genICPSFromFile generates an IstioControlPlaneSpec for a spec file
+func genICPSFromFile(filename string, force bool) *v1alpha2.IstioControlPlaneSpec {
+	_, overlayICPS, err := genICPS(filename, "", "", force, l)
```

Comment:
This seems like overkill for a function. Anyway you want to pass an error up rather than logging here, which is what the function you're wrapping is doing.

Reply:
Yes, will add error for all those functions. 
The original thought was to keep the upgrade() neat and easy-to-read.

Reply:
Func inlined.